### PR TITLE
Carry save multiplier

### LINKED ISSUE #10: Pipelined integer multiply
### Motivation

Integer multiplication in a pipeline is a nice component for the library

### Desired solution

A simple pipelined integer multiplier

### Alternatives considered

_No response_

### Additional details

_No response_

### PR DESCRIPTION
**Pull request by quekyj**
## Description & Motivation

Add carry save multiplier module.

## Related Issue(s)

Fix #10 

## Testing

Added unit test in the test folder

## Backwards-compatibility

> Is this a breaking change that will not be backwards-compatible? If yes, how so?

No

## Documentation

> Does the change require any updates to documentation? If so, where? Are they included?

Yes

### CODE REVIEW

File: lib/src/carry_save_multiplier.dart [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+/// A datatype class that hold result from [FullAdder].
+class FullAdderResult {
+  /// The output sum from full adder.
+  final sum = Logic(name: 'sum');
+
+  /// The output carry-out from full adder.
+  final cOut = Logic(name: 'c_out');
+}
+
+/// A [FullAdder] module that can perform simple adder. The full adder output
```
Comment (mkorbel1) [inline lib/src/carry_save_multiplier.dart:22]: Doc comment is a little confusing.  Let's describe it more specifically, something more like
```dart
/// A simple full-adder with inputs [a] and [b] to be added with a [carryIn].
```

File: lib/src/carry_save_multiplier.dart [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+/// can be defined in the truth table.
+class FullAdder extends Module {
```
Comment (mkorbel1) [inline lib/src/carry_save_multiplier.dart:24]: A full adder is a good component all by itself, we should include it in a separate file?

File: lib/src/carry_save_multiplier.dart [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+// Implementation of pipeline multiplier module.
+//
+// 2023 February 21
+// Author: Max Korbel <max.korbel@intel.com>
+//
+
+import 'package:rohd/rohd.dart';
+
+/// A datatype class that hold result from [FullAdder].
+class FullAdderResult {
```
Comment (mkorbel1) [inline lib/src/carry_save_multiplier.dart:14]: Let's ditch this `FullAdderResult` object and just have `sum` and `cOut` directly as outputs (`get`ters) of the `FullAdder` module.

File: lib/src/carry_save_multiplier.dart [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+    b = addInput('b', b, width: b.width);
+    carryIn = addInput('carry_in', carryIn, width: carryIn.width);
+
+    final carryOut = addOutput('carry_out');
+    final sum = addOutput('sum');
+
+    final and1 = carryIn & (a ^ b);
+    final and2 = b & a;
+
+    Combinational([
```
Comment (mkorbel1) [inline lib/src/carry_save_multiplier.dart:48]: Use assigns `<=` isntead of a ` Combinational` to make things cleaner.

File: lib/src/carry_save_multiplier.dart [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+      carryOut < and1 | and2,
+    ]);
+
+    fullAdderResult.sum <= output('sum');
+    fullAdderResult.cOut <= output('carry_out');
+  }
+}
+
+/// An [NBitAdder] that perform addition.
+class NBitAdder extends Module {
```
Comment (mkorbel1) [inline lib/src/carry_save_multiplier.dart:59]: This is a good component by itself, put it in another file?

**Review by mkorbel1 — CHANGES_REQUESTED** [review]

File: lib/src/carry_save_multiplier.dart [hunk 6]
```diff
@@ -4,29 +4,19 @@
...
 class FullAdder extends Module {
-  /// The variable that hold [FullAdder] results.
-  final fullAdderResult = FullAdderResult();
+  /// The result [sum] from [FullAdder].
+  Logic get sum => output('sum');
 
-  /// The results returned from [FullAdder].
-  FullAdderResult get fullAdderRes => fullAdderResult;
+  /// The result [cOut] from [FullAdder].
+  Logic get cOut => output('carry_out');
```
Comment (mkorbel1) [inline lib/src/carry_save_multiplier.dart:19]: Either call this `carryOut` or call the input `cIn`, should be consistent.

File: lib/src/carry_save_multiplier.dart [hunk 7]
```diff
@@ -4,29 +4,19 @@
...
+/// A simple full-adder with inputs `a` and `b` to be added with a `carryIn`.
 class FullAdder extends Module {
-  /// The variable that hold [FullAdder] results.
-  final fullAdderResult = FullAdderResult();
+  /// The result [sum] from [FullAdder].
+  Logic get sum => output('sum');
 
-  /// The results returned from [FullAdder].
-  FullAdderResult get fullAdderRes => fullAdderResult;
+  /// The result [cOut] from [FullAdder].
```
Comment (mkorbel1) [inline lib/src/carry_save_multiplier.dart:18]: Clarify what this is, it's the carry bit.  Calling it `[cOut]` doesnt' help in the docs.

**Review by mkorbel1 — CHANGES_REQUESTED** [review]

File: doc/multiplier.md [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+
+ROHD HCL provides a multiplier module to get the product from a list of logic. As of now, we have
+
+- [Carry Save Multiplier](#carry-save-multiplier)
+
+## Carry Save Multiplier
+
+Carry save multiplier is a digital circuit used for performing multiplication operations. It is particularly useful in applications that require high speed multip0lication, such as digital signal processing.
+
+The [`CarrySaveMultiplier`](https://intel.github.io/rohd-hcl/rohd_hcl/CarrySaveMultiplier-class.html) module in ROHD-HCl accept four inputs: the clock `clk` signal, reset `reset` signal, a List of Logic()'s `toMultiply` and the name of the module `name`. Note that `toMultiply` must have inputs of two and width of the inputs must be the same or `RohdHclException()` will be thrown.
```
Comment (mkorbel1) [inline doc/multiplier.md:11]: Typo "ROHD-HCl" should be "ROHD-HCL"

Comment (mkorbel1) [inline doc/multiplier.md:11]: Instead of "Logic()", say `Logic` (in ticks).  This applies in the adder doc also.  There's no need to add `()` after class names (e.g. in `RohdHclException()`)

File: lib/src/adder.dart [hunk 9]
```diff
@@ -0,0 +1,55 @@
+// Copyright (C) 2023 Intel Corporation
+// SPDX-License-Identifier: BSD-3-Clause
+//
+// carry_save_multiplier.dart
```
Comment (mkorbel1) [inline lib/src/adder.dart:4]: Check the names of files and file descriptions, copy-paste error in multiple files.

File: lib/src/ripple_carry_adder.dart [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+    final b = addInput('b', toSum[1], width: toSum[1].width);
+    carry = addInput('carry_in', carry, width: carry.width);
+
+    if (a.width != b.width) {
+      throw RohdHclException('inputs in toSum should have same width.');
+    }
+
+    for (var i = 0; i < a.width; i++) {
+      FullAdder? fullAdder;
+      fullAdder = FullAdder(a: a[i], b: b[i], carryIn: carry);
```
Comment (mkorbel1) [inline lib/src/ripple_carry_adder.dart:45]: Why not just `final fullAdder = ...`?

File: lib/src/adder.dart [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+
+import 'package:rohd/rohd.dart';
+
+/// An abstract class for all adder module.
+abstract class Adder extends Module {
+  /// The List of inputs to sum. Length of inputs must be two.
+  List<Logic> toSum;
+
+  /// The addition results [sum].
+  List<Logic> get sum;
```
Comment (mkorbel1) [inline lib/src/adder.dart:19]: The output of an adder should be a `Logic`, not a `List<Logic>`.  Swizzle it in implementations.

File: lib/src/carry_save_mutiplier.dart [hunk 12]
```diff
@@ -0,0 +1,126 @@
...
+/// A binary multiplier using carry-save technique with pipelineing.
+///
+/// Reduces calculation time and complexity by employing the carry save
+/// multiplier, which splits numbers into smaller components and performs
+/// partial multiplications on each component separately, storing the results
+/// in a compact form. The pipeline enhances performance by breaking down the
+/// multiplication process into sequential stages, allowing for concurrent
+/// execution of multiple operations.
+///
+/// The latency of the carry save multiplier is the sum of the two inputs width
```
Comment (mkorbel1) [inline lib/src/carry_save_mutiplier.dart:23]: Can you add a getter that calculates the latency of the multiplier, also?

File: lib/src/ripple_carry_adder.dart [hunk 13]
```diff
@@ -0,0 +1,53 @@
...
+  RippleCarryAdder({required super.toSum, super.name = 'ripple_carry_adder'}) {
+    Logic carry = Const(0);
+
+    if (toSum.length != 2) {
+      throw RohdHclException('Length of toSum must be two.');
+    }
+
+    final a = addInput('a', toSum[0], width: toSum[0].width);
+    final b = addInput('b', toSum[1], width: toSum[1].width);
+    carry = addInput('carry_in', carry, width: carry.width);
```
Comment (mkorbel1) [inline lib/src/ripple_carry_adder.dart:37]: This carry is not an input to the module, so delete this line.

File: test/ripple_carry_adder_test.dart [hunk 14]
```diff
@@ -0,0 +1,55 @@
...
+    test('should throw exception if toSum length is not two.', () {
+      final a = Logic(name: 'a', width: 8);
+      final b = Logic(name: 'b', width: 8);
+      final c = Logic(name: 'c', width: 8);
+
+      expect(() => RippleCarryAdder(toSum: [a, b, c]),
+          throwsA(const TypeMatcher<RohdHclException>()));
+    });
+
+    test('should return correct value for ripple carry adder.', () {
```
Comment (mkorbel1) [inline test/ripple_carry_adder_test.dart:40]: Can we add some corner cases? E.g.

- 0+0
- ?+0
- Carry bit non-zero
- Overflow
- A bunch of random numbers in a loop

File: lib/src/carry_save_mutiplier.dart [hunk 15]
```diff
@@ -0,0 +1,126 @@
...
+
+  /// The list pf carry from every pipeline stages.
+  late final List<Logic> _carry;
+
+  /// The final product of the multiplier module.
+  @override
+  Logic get product => output('product');
+
+  /// The pipeline for [CarrySaveMultiplier].
+  late final Pipeline pipeline;
```
Comment (mkorbel1) [inline lib/src/carry_save_mutiplier.dart:37]: pipeline should be private

File: lib/src/carry_save_mutiplier.dart [hunk 16]
```diff
@@ -0,0 +1,126 @@
...
+
+  /// Construct a [CarrySaveMultiplier] that multiply two list of logic in
+  /// [toMultiply].
+  CarrySaveMultiplier(Logic clk, Logic reset,
```
Comment (mkorbel1) [inline lib/src/carry_save_mutiplier.dart:41]: let's make `clk` and `reset` named arguments so it's not ambiguous and to avoid people putting `a` and `b` there.

File: test/carry_save_multiplier_test.dart [hunk 17]
```diff
@@ -0,0 +1,82 @@
...
+    unawaited(Simulator.run());
+
+    Future<void> waitCycles(int numCycles) async {
+      for (var i = 0; i < numCycles; i++) {
+        await clk.nextPosedge;
+      }
+    }
+
+    final inputs = List.generate(
+        10, (index) => List.generate(2, (index) => Random().nextInt(100) + 1));
```
Comment (mkorbel1) [inline test/carry_save_multiplier_test.dart:67]: Instead of 100, make it fill the width (`1<<widthLength`)

File: test/carry_save_multiplier_test.dart [hunk 18]
```diff
@@ -0,0 +1,82 @@
...
+
+    for (final input in inputs) {
+      a.put(input[0]);
+      b.put(input[1]);
+
+      await waitCycles(widthLength * 2).then((value) {
```
Comment (mkorbel1) [inline test/carry_save_multiplier_test.dart:73]: I think you want `unawaited`, not `await` on this line, so that they are going in parallel

**Review by mkorbel1 — CHANGES_REQUESTED** [review]
This is looking great

File: doc/multiplier.md [hunk 19]
```diff
@@ -0,0 +1,24 @@
...
+An example is shown below to multiply two inputs of signals that have 8-bits of width.
+
+```dart
+const dataWidth = 8;
+final clk = SimpleClockGenerator(10).clk;
+final reset = Logic(name: 'reset');
+final a = Logic(name: 'a', width: dataWidth);
+final b = Logic(name: 'b', width: dataWidth);
+
+final multiply = CarrySaveMultiplier(clk: clk, reset: reset, a, b, name: 'csm_module');
```
Comment (mkorbel1) [inline doc/multiplier.md:22]: In this example you should show getting the `product` from it as well?

File: doc/adder.md [hunk 20]
```diff
@@ -0,0 +1,24 @@
...
+
+- [Ripple Carry Adder](#ripple-carry-adder)
+
+## Ripple Carry Adder
+
+An ripple carry adder is a digital circuit used for binary addition. It consists of a series of full adders connected in a chain, with the carry output of each adder linked to the carry input of the next one. Starting from the least significant bit (LSB) to most significant bit (MSB), the adder sequentially adds corresponding bits of two binary numbers.
+
+The [`RippleCarryAdder`](https://intel.github.io/rohd-hcl/rohd_hcl/RippleCarryAdder-class.html) module in ROHD-HCL accept input parameters `Logic`s' a and b as the input pin and the name of the module `name`. Note that the width of the inputs must be the same or `RohdHclException` will be thrown.
+
+An example is shown below to multiply two inputs of signals that have 8-bits of width.
```
Comment (mkorbel1) [inline doc/adder.md:13]: Typo, you say `multiply` here but this is an adder

File: lib/src/adder.dart [hunk 21]
```diff
@@ -0,0 +1,58 @@
...
+// 2023 June 1
+// Author: Yao Jing Quek <yao.jing.quek@intel.com>
+//
+
+import 'package:rohd/rohd.dart';
+
+/// An abstract class for all adder module.
+abstract class Adder extends Module {
+  /// The input to the adder pin [a].
+  Logic a;
```
Comment (mkorbel1) [inline lib/src/adder.dart:16]: the `input` signals of the class should _not_ be publicly visible.  they should also probably be `final` or at least `late final`.  At least we should use `@protected` if we don't find a nice way to make them private.

same applies to the multiplier

File: lib/src/adder.dart [hunk 22]
```diff
@@ -0,0 +1,58 @@
...
+
+  /// The input to the adder pin [b].
+  Logic b;
+
+  /// The addition results [sum].
+  Logic get sum;
+
+  /// Takes in input [a] and input [b] and return the [sum] of the addition
+  /// result.
+  Adder(this.a, this.b, {super.name});
```
Comment (mkorbel1) [inline lib/src/adder.dart:26]: You could take care of the `addInput`s, width checks, etc. in here since they should be common across all adder implementations.

same applies to the multiplier

File: test/ripple_carry_adder_test.dart (hunk 14 above)
Comment (mkorbel1) [reply test/ripple_carry_adder_test.dart]: missing the random numbers in a loop test?

**Review by mkorbel1 — CHANGES_REQUESTED** [review]

File: test/ripple_carry_adder_test.dart (hunk 14 above)
Comment (quekyj) [reply test/ripple_carry_adder_test.dart]:
```dart
test('should return correct value when random numbers is given.', () async {
```

The test is here.

File: lib/src/adder.dart [hunk 23]
```diff
@@ -0,0 +1,66 @@
...
+
+  /// Constructs a [FullAdder] with value [a], [b] and [carryIn] based on
+  /// full adder truth table.
+  FullAdder({
+    required Logic a,
+    required Logic b,
+    required Logic carryIn,
+    super.name = 'full_adder',
+  }) {
+    a = addInput('a', a, width: a.width);
```
Comment (mkorbel1) [inline lib/src/adder.dart:53]: Move these `addInput`s for `a` and `b` into `Adder`, since you made the protected variables there.  That will also probably force them to be `late final` in `Adder`.

File: lib/src/carry_save_mutiplier.dart [hunk 24]
```diff
@@ -0,0 +1,125 @@
...
+
+  /// The pipeline for [CarrySaveMultiplier].
+  late final Pipeline _pipeline;
+
+  /// Construct a [CarrySaveMultiplier] that multiply input [a] and input [b].
+  CarrySaveMultiplier(super.a, super.b,
+      {required Logic clk,
+      required Logic reset,
+      super.name = 'carry_save_multiplier'}) {
+    final portA = addInput('a', a, width: a.width);
```
Comment (mkorbel1) [inline lib/src/carry_save_mutiplier.dart:47]: Same here, these should be moved into `Multiplier` and then use `a` and `b` inherited from there

File: pubspec.yaml [hunk 25]
```diff
@@ -10,6 +10,7 @@ environment:
   sdk: '>=2.18.0 <3.0.0'
 
 dependencies:
+  meta: ^1.9.1
```
Comment (mkorbel1) [inline pubspec.yaml:13]: New dependency, will throw this through the review process

**Review by mkorbel1 — CHANGES_REQUESTED** [review]

File: lib/src/adder.dart (hunk 23 above)
Comment (quekyj) [reply lib/src/adder.dart:53]: Its seems like I need to removed the final.

```dart
/// An abstract class for all adder module.
abstract class Adder extends Module {
  /// The input to the adder pin [a].
  @protected
  late final Logic a;

  /// The input to the adder pin [b].
  @protected
  late final Logic b;

  /// The addition results [sum].
  Logic get sum;

  /// Takes in input [a] and input [b] and return the [sum] of the addition
  /// result. The width of input [a] and [b] must be the same.
  Adder(this.a, this.b, {super.name}) {
    if (a.width != b.width) {
      throw RohdHclException('inputs of a and b should have same width.');
    }
    a = addInput('a', a, width: a.width);
    b = addInput('b', b, width: b.width);
  }
}
```

Since `a` and `b` will be reinitialized.

Comment (mkorbel1) [reply lib/src/adder.dart:53]: You should make it `late final`

Comment (mkorbel1) [reply lib/src/adder.dart:53]: And don't make the inputs `this.a` and `this.b`, just `Logic a` and `Logic b`, then `this.a = addInput...`

**Review by mkorbel1 — APPROVED** [review]
Nice work! Looks good!
